lib/srv/db/common/permissions/import_rules.go

```diff
@@ -0,0 +1,141 @@
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package permissions
+
+import (
+	"regexp"
+	"sort"
+
+	dbobjectv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/dbobject/v1"
+	dbobjectimportrulev1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/dbobjectimportrule/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/types/label"
+	"github.com/gravitational/teleport/api/utils"
+	"github.com/gravitational/teleport/lib/services"
+	utils2 "github.com/gravitational/teleport/lib/utils"
+)
+
+// ApplyDatabaseObjectImportRules applies the given set of rules onto a set of objects coming from a same database.
+// Returns a fresh copy of a subset of supplied objects, filtered and modified.
+// For the object to be returned, it must match at least one rule.
+// The modification consists of application of extra labels, per matching mappings.
+func ApplyDatabaseObjectImportRules(rules []*dbobjectimportrulev1.DatabaseObjectImportRule, database types.Database, objs []*dbobjectv1.DatabaseObject) []*dbobjectv1.DatabaseObject {
+	// sort: rules with higher priorities are applied last.
+	sort.Slice(rules, func(i, j int) bool {
+		return rules[i].Spec.Priority < rules[j].Spec.Priority
+	})
+
+	// filter rules: keep those with matching labels
+	// we only need mappings from the rules, so extract those.
+	var mappings []*dbobjectimportrulev1.DatabaseObjectImportRuleMapping
+	for _, rule := range rules {
+		dbLabels := make(types.Labels)
+		mapLabel := label.ToMap(rule.Spec.GetDatabaseLabels())
+		for k, v := range mapLabel {
+			dbLabels[k] = v
+		}
+		if ok, _, _ := services.MatchLabels(dbLabels, database.GetAllLabels()); ok {
+			mappings = append(mappings, rule.Spec.Mappings...)
+		}
+	}
+
+	// anything to do?
+	if len(mappings) == 0 {
+		return nil
+	}
+
+	var out []*dbobjectv1.DatabaseObject
+
+	// find all objects that match any of the rules
+	for _, obj := range objs {
+		var objClone *dbobjectv1.DatabaseObject
+
+		// apply each mapping in order.
+		for _, mapping := range mappings {
+			// the matching is applied to the object spec; existing object labels does not matter
+			if !databaseObjectScopeMatch(mapping.GetScope(), obj.GetSpec()) {
+				continue
+			}
+			if databaseObjectImportMatch(mapping.GetMatch(), obj.GetSpec()) {
+				if objClone == nil {
+					objClone = utils.CloneProtoMsg(obj)
+				}
+
+				// mapping applies additional labels
+				labels := objClone.Metadata.Labels
+				if labels == nil {
+					labels = map[string]string{}
+				}
+				for k, v := range mapping.AddLabels {
+					labels[k] = v
+				}
+				objClone.Metadata.Labels = labels
+			}
+		}
+
+		if objClone != nil {
+			out = append(out, objClone)
+		}
+	}
+
+	return out
+}
+
+func matchPattern(pattern, value string) bool {
+	matched, _ := regexp.MatchString("^"+utils2.GlobToRegexp(pattern)+"$", value)
+	return matched
+}
+
+func matchAny(patterns []string, value string) bool {
```

Comment:
I tried it, but turns out it does the inverse of what we need here: `RegexMatchesAny` is for single pattern against multiple values, whereas here I need to try multiple patterns against single value.

Still, I've decided to use `CompileExpression` in `matchPattern` instead of custom string concatenation.